### ISSUE TITLE
NBTIngredient.of() does not save NBT

### ISSUE THREAD

**Issue by IchHabeHunger54** [MinecraftForge/MinecraftForge#8458]
**Minecraft Version:** 1.18.1

**Forge Version:** 39.0.66, pretty sure every version prior to that

**Description of issue:**

NBTIngredient does not override the of() method. Instead, the normal Ingredient.of() implementation is called, which specifically creates a new Ingredient (not NBTIngredient). As a result, the behavior saving the stack, as the NBTIngredient constructor would, is never called, effectively having an NBTIngredient without any NBT.

**Comment by pupnewfster** [comment]
This is the normal behavior of static methods in java?

**Comment by IchHabeHunger54** [comment]
It sure is, but that leads to the actual issue, being that NBTIngredient.of() returns a normal Ingredient which ignores NBT, which I cannot believe is intended behavior.

**Comment by diesieben07** [comment]
NBTIngredient does not have an `of` method. You are calling `Ingredient.of`.

**Comment by SizableShrimp** [comment]
This is simply how static methods in superclasses work. I would consider using `NBTIngredient#of` "wrong" because you are referencing it from a subclass rather than the actual class that it is declared in. At least in IntelliJ, this emits a warning stating this fact. I don't see this as a bug.

**Comment by IchHabeHunger54** [comment]
But what should I use then to create a new NBTIngredient? Its constructor is protected, I'd need to create a subclass making the constructor public (which is my current solution, but it's rather ugly).

**Comment by desht** [comment]
Subclassing is the best approach given the circumstances.  I have an inner class like this in my recipe datagen class:
```java
    // this wrapper is due to lack of any public constructor method for IngredientNBT
    private static class IngredientNBTWrapper extends NBTIngredient {
        IngredientNBTWrapper(ItemStack stack) {
            super(stack);
        }

        static IngredientNBTWrapper fromItemStack(ItemStack stack) {
            return new IngredientNBTWrapper(stack);
        }
    }
```
and then just use `IngredientNBTWrapper.fromItemStack(stack)` as needed.

**Comment by IchHabeHunger54** [comment]
Yes, so do I. I just wonder why this isn't in Forge already.

**Comment by desht** [comment]
Probably simply because no one's ever gotten around to PR'ing in a static factory method, since workarounds are easy.

Another approach is to use an anonymous subclass:
```java
NBTIngredient n = new NBTIngredient(stack) { };
```

**Comment by sciwhiz12** [comment]
Fixed by #8517 (released as 1.18.2-40.0.30) with the addition of the `NBTIngredient.of` helper which properly constructs an `NBTIngredient`, avoiding the call `NBTIngredient.of` from actually calling `Ingredient.of`.
